### ISSUE TITLE
requirements.lock does not get updated on helm dep up

### ISSUE THREAD

**Issue by azazi-sa** [helm/helm#2731]
if there is no change in `requirements.yaml` file but SemVar results into different lock file, `helm dep up` does not replace and existing `requirements.lock` because `requirements.yaml` digest has not changed.

How to reproduce:

We have the following `requirements.yaml` file for a chart `parent1`.

```
dependencies:
- name: child1
  version: ">=1.0.0"
  repository: "http://127.0.0.1:8879/"
- name: child2
  version: ">=1.0.0"
  repository: "http://127.0.0.1:8879/"
```

When we run dep up for the first time it generates following lock file (assuming that we have `child1-1.0.0` and `child2-1.0.0` charts at that point)

```
dependencies:
- name: child1
  repository: http://127.0.0.1:8879/
  version: 1.0.0
- name: child2
  repository: http://127.0.0.1:8879/
  version: 1.0.0
digest: sha256:19b7f91a0fdfc44e14a86e7b3bbfbce2014e063bdc5949c2ad6a5edb113c3b77
generated: 2017-07-27T12:00:08.036838191+05:30

```

Now if we release a new version of `child1` (say 1.1.0) we don't need to update our `requirements.yaml` file but we expect `helm dep up` to update `requirements.lock` file accordingly. 

Instead, `helm dep up` updates `/charts` folder with `child1-1.1.0` but does not update `requirements.lock` file on disk.

When we looked into code, we figured that `requirements.lock` only get updated if `requirements.yaml` gets changed. 

```
// Hash requirements.yaml
hash, err := resolver.HashReq(req)
if err != nil {
         return err
}

// Now we need to find out which version of a chart best satisfies the
// requirements the requirements.yaml
lock, err := m.resolve(req, repoNames, hash)
if err != nil {
	return err
}

// If the lock file hasn't changed, don't write a new one.
oldLock, err := chartutil.LoadRequirementsLock(c)
if err == nil && oldLock.Digest == lock.Digest {
	return nil
}
```

This results into inconsistent state of having different charts in `/charts` directory and `requirements.lock` file. Only work around, currently, is to delete `requirements.lock` file when we do `helm dep up`

**Comment by jascott1** [comment]
Hi @azazi-sa,

Sorry but the command may be counter-intuitive. The `helm dep up` command exists to realize the `requirements.yaml` by creating/updating the lock file and downloading charts to fulfill that lock file. It does not update the requirements.yaml or lock file from charts on disk.

The `requirements.yaml` is intended to represent the desired state and allows for SemVer compatible chart versions (e.g. ">1.5.0") and the `requirements.lock` is a concrete version of that (perhaps containing version 1.5.1 which satisfies the SemVer). 

I do think that chart version manipulation would be a very useful feature and I am planning to author a proposal in the near future.  Hopefully you can comment on that proposal (or create your own) to ensure your use case gets covered (I will make sure its linked from here).

Also just a thought but in the short term you might implement something as a Helm plugin using the language of your choice.

**Comment by rowleyaj** [comment]
Hey @jascott1,

I've been using 2.5.1 today and testing out dependency management. I'm experiencing similar problems to @azazi-sa.

My understanding is that `helm dep build` does one of two things dependant on the presence of a `requirements.lock` file. If the file is there is uses it to pull the version specified, if not is uses `requirements.yaml` to pull the version specified.

`helm dep update` is intended to be the command to update dependencies, by creating/updating the lockfile.

When I run update it doesn't seem to be updating the lock file, but does pull the correct chart version and place it into `/charts`, as @azazi-sa said.

I've created a really simple example repo that should show what was happening:
https://github.com/rowleyaj/helm-update-debug/commits/master

I've committed explaining what commands were ran, I didn't want to commit the .tgz files, so there are some empty commits where nothing changed on disk - the expected behaviour would have been that the lock file changed when update was ran.

```
Commits:
Run helm dep update charts/chartb
Creates charts/chartb/charts/charta-0.2.0.tgz

Run helm dep build charts/chartb
Gives error:
Error: can't get a valid version for dependency charta
This is expected

Update charta to 0.2.0

Run helm dep build charts/chartb
Creates charts/chartb/charts/charta-0.1.0.tgz

Commit basic charts
```

**Comment by jascott1** [comment]
HI @rowleyaj 

I cannot reproduce this with Helm 2.5.1 or 2.6.2. In my manual testing the requirements.lock gets updated. Could there be anything else going on?

**Comment by rowleyaj** [comment]
Thanks for looking into it @jascott1. The example in the repo was using local file based charts, could that make a difference? Could you share how you tested this and I can try and get it working from that.

If I can produce both a working and non-working example then I might be able to work out whats going wrong.

**Comment by jascott1** [comment]
@rowleyaj I used the chart you linked and followed the same steps except my lock file gets the updated 0.2.0 version. Im at a loss as to what could be different in your env.

**Comment by bacongobbler** [comment]
I'm going to close this support ticket due to inactivity, but please re-open if this still needs to be addressed. I'd suggest trying the latest version of Helm and see if that fixes things first before re-opening so we can try again. Thanks!

**Comment by jgiles** [comment]
I'm also seeing this issue, using Helm 2.7.2 installed on OSX using `brew install kubernetes-helm`.

I am not using local `file://` charts. As suggested above, my workaround is to delete the lock file before running `helm dep up`

**Comment by munkyboy** [comment]
I'm running into this problem as well. using helm 2.8.2 and S3 as a repo via https://github.com/hypnoglow/helm-s3

when I run `helm dep up`, new sub-chart tarballs appear but `requirements.lock` is not updated.

**Comment by VojtechStep** [comment]
I'm experiencing the same issue (new charts get downloaded but `requirements.lock` stays the same) on Windows with Helm v2.10.0

**Comment by dbluxo** [comment]
Same with Helm v2.11.0

**Comment by hpurmann** [comment]
@bacongobbler Can you please re-open? We just ran into this issue as well and as others report as well, this seems to still be a problem for more people.

**Comment by jotadrilo** [comment]
It seems like this issue is duplicated (https://github.com/helm/helm/issues/4011)

**Comment by carrodher** [comment]
The same issue also using Helm v2.11.0.

**Comment by bacongobbler** [comment]
Fixed in Helm 3 via #5874. Closing!
